{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "365c22b2-edb2-4bfb-a1a5-9fae90e981ba",
   "metadata": {},
   "source": [
    "#### Question 1\n",
    "In a practical, we saw the scs function (copied below along with overlap) for finding the shortest common superstring of a set of strings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f58d3e0-9bab-489d-8b62-e2262508318c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap(a, b, min_length=3):\n",
    "    \"\"\" Return length of longest suffix of 'a' matching\n",
    "        a prefix of 'b' that is at least 'min_length'\n",
    "        characters long.  If no such overlap exists,\n",
    "        return 0. \"\"\"\n",
    "    start = 0  # start all the way at the left\n",
    "    while True:\n",
    "        start = a.find(b[:min_length], start)  # look for b's suffx in a\n",
    "        if start == -1:  # no more occurrences to right\n",
    "            return 0\n",
    "        # found occurrence; check for full suffix/prefix match\n",
    "        if b.startswith(a[start:]):\n",
    "            return len(a)-start\n",
    "        start += 1  # move just past previous match\n",
    "\n",
    "import itertools\n",
    "\n",
    "def scs(ss):\n",
    "    \"\"\" Returns shortest common superstring of given\n",
    "        strings, which must be the same length \"\"\"\n",
    "    shortest_sup = None\n",
    "    for ssperm in itertools.permutations(ss):\n",
    "        sup = ssperm[0]  # superstring starts as first string\n",
    "        for i in range(len(ss)-1):\n",
    "            # overlap adjacent strings A and B in the permutation\n",
    "            olen = overlap(ssperm[i], ssperm[i+1], min_length=1)\n",
    "            # add non-overlapping portion of B to superstring\n",
    "            sup += ssperm[i+1][olen:]\n",
    "        if shortest_sup is None or len(sup) < len(shortest_sup):\n",
    "            shortest_sup = sup  # found shorter superstring\n",
    "    return shortest_sup  # return shortest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce05ab0d-7341-4f89-9be9-be7d7aeb2d07",
   "metadata": {},
   "source": [
    "It's possible for there to be multiple different shortest common superstrings for the same set of input strings. Consider the input strings:\n",
    "\n",
    "- **ABC**\n",
    "- **BCA**\n",
    "- **CAB**\n",
    "\n",
    "One shortest common superstring is **ABCAB**, but another is **BCABC**, and another is **CABCA**.\n",
    "\n",
    "Now, what is the length of the shortest common superstring of the following strings?\n",
    "\n",
    "- **CCT**\n",
    "- **CTT**\n",
    "- **TGC**\n",
    "- **TGG**\n",
    "- **GAT**\n",
    "- **ATT**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b28b1df-9dae-4002-9179-c713650338e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "strings = ['CCT', 'CTT', 'TGC','TGG','GAT','ATT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ba3b1d03-aad7-49fe-b803-c173741988e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(scs(strings))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39f83353-c893-4699-a6d7-13b29532a8f0",
   "metadata": {},
   "source": [
    "#### Question 2\n",
    "How many different shortest common superstrings are there for the input strings given in the previous question?\n",
    "\n",
    "Hint 1: You can modify the scs function to keep track of this. \n",
    "\n",
    "Hint 2: You can look at these examples to double-check that your modified scs is working as expected. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "52cd50a6-1661-4a7d-b316-f31214c9571c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import permutations\n",
    "\n",
    "def scs_list(strings):\n",
    "    \"\"\"Return all shortest common superstrings from all permutations of input strings.\"\"\"\n",
    "    shortest_sup = None\n",
    "    result = set()\n",
    "    \n",
    "    for ssperm in permutations(strings):\n",
    "        sup = ssperm[0]\n",
    "        for i in range(len(strings)-1):\n",
    "            olen = overlap(sup, ssperm[i+1], min_length=1)\n",
    "            sup += ssperm[i+1][olen:]\n",
    "        \n",
    "        if shortest_sup is None or len(sup) < len(shortest_sup):\n",
    "            shortest_sup = sup\n",
    "            result = {sup}\n",
    "        elif len(sup) == len(shortest_sup):\n",
    "            result.add(sup)\n",
    "    \n",
    "    return sorted(result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d23ffad-16c9-4577-86e8-843071ef9e2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(scs_list(strings))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6da817dc-dfb4-4d56-a333-8e3d19a89edf",
   "metadata": {},
   "source": [
    "#### Question 3\n",
    "\n",
    "All the reads are the same length (100 bases) and are exact copies of substrings from the forward strand of the virus genome.  You don't have to worry about sequencing errors, ploidy, or reads coming from the reverse strand.\n",
    "\n",
    "Assemble these reads using one of the approaches discussed, such as greedy shortest common superstring.  Since there are many reads, you might consider ways to make the algorithm faster, such as the one discussed in the programming assignment in the previous module.\n",
    "\n",
    "How many As (Adenines) are there in the full, assembled genome?\n",
    "\n",
    "Hint: the virus genome you are assembling is exactly 15,894 bases long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "40b71836-c90a-40b5-b144-ae1ac9b16efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip()  # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "sequences, qualities = readFastq('ads1_week4_reads.fq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1bfe68bc-9af1-4503-a4f2-ba6e70204cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "from collections import defaultdict\n",
    "\n",
    "def compute_all_overlaps(reads, k):\n",
    "    \"\"\" Precompute all overlaps of length >= k between different reads \"\"\"\n",
    "    overlaps = {}\n",
    "    for a, b in itertools.permutations(reads, 2):\n",
    "        olen = overlap(a, b, min_length=k)\n",
    "        if olen > 0:\n",
    "            overlaps[(a, b)] = olen\n",
    "    return overlaps\n",
    "\n",
    "def greedy_scs_precomputed(reads, k):\n",
    "    reads = reads.copy()\n",
    "    overlaps = compute_all_overlaps(reads, k)\n",
    "\n",
    "    while True:\n",
    "        if not overlaps:\n",
    "            break\n",
    "\n",
    "        # Find pair with max overlap\n",
    "        (read_a, read_b), olen = max(overlaps.items(), key=lambda x: x[1])\n",
    "        reads.remove(read_a)\n",
    "        reads.remove(read_b)\n",
    "        merged = read_a + read_b[olen:]\n",
    "        reads.append(merged)\n",
    "\n",
    "        # Recompute overlaps involving new merged read\n",
    "        overlaps = {\n",
    "            (a, b): o for (a, b), o in overlaps.items()\n",
    "            if a in reads and b in reads and a != b\n",
    "        }\n",
    "        for r in reads:\n",
    "            if r != merged:\n",
    "                o1 = overlap(r, merged, k)\n",
    "                o2 = overlap(merged, r, k)\n",
    "                if o1 > 0:\n",
    "                    overlaps[(r, merged)] = o1\n",
    "                if o2 > 0:\n",
    "                    overlaps[(merged, r)] = o2\n",
    "\n",
    "    return ''.join(reads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "88508920-d6ab-49fa-b9b3-b952c89c3f52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome assembled!\n",
      "A count: 4633\n",
      "T count: 3723\n",
      "CPU times: total: 2min 35s\n",
      "Wall time: 2min 36s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for k in range(30, 1, -1):  # 30 is enough, no need to start at 100\n",
    "    genome = greedy_scs_precomputed(sequences, k)\n",
    "    if len(genome) == 15894:\n",
    "        print(\"Genome assembled!\")\n",
    "        print(\"A count:\", genome.count('A'))\n",
    "        print(\"T count:\", genome.count('T'))\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a322a15b-cf05-4a26-876a-7a37f92075ea",
   "metadata": {},
   "source": [
    "#### Question 4\n",
    "\n",
    "How many Ts are there in the full, assembled genome from the previous question?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95e4a7c1-73d5-4b36-81b8-1d88f09db8ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
