{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b11e55f3-7b30-40a2-84f2-15a31cafad04",
   "metadata": {},
   "source": [
    "## Homework"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "807b1bdd-e202-467f-9aa8-697156c06427",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bm_preproc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f27263a9-b61f-4151-be08-78463984ed68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def boyer_moore(p, p_bm, t):\n",
    "    \"\"\" Do Boyer-Moore matching. p=pattern, t=text,\n",
    "        p_bm=BoyerMoore object for p \"\"\"\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    alignments = 0\n",
    "    comparisons = 0\n",
    "    while i < len(t) - len(p) + 1:\n",
    "        alignments += 1\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            comparisons += 1\n",
    "            if p[j] != t[i+j]:\n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "    return occurrences, alignments, comparisons\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ae6f8c9-c2a8-49e7-bfbe-9ac3ebebfccd",
   "metadata": {},
   "source": [
    "#### Question 1\n",
    "Boyer-Moore: How many alignments are skipped by the bad character rule for this alignment?\n",
    "\n",
    "Note: the number of skips is one less than the number of positions P shifts by.  That is, if the pattern shifts by 2 positions, that's 1 alignment skipped.\n",
    "\n",
    "Also note: the question is asking only about the alignment shown.  Do not consider any other alignments of P to T in your answer.\n",
    "\n",
    "T: GGCTATAATGCGTA\n",
    "\n",
    "P:  TAATAAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "887c12c2-15c4-4c5f-aedf-8d790463d508",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n"
     ]
    }
   ],
   "source": [
    "T = 'GGCTATAATGCGTA'\n",
    "P = 'TAATAAA'\n",
    "P_bm = BoyerMoore(P)\n",
    "skip = P_bm.bad_character_rule(4, 'G')\n",
    "skipped_alignments = skip - 1\n",
    "print(skipped_alignments)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e48df205-0e87-44fa-a98d-8b0cb1f9c7b8",
   "metadata": {},
   "source": [
    "#### Question 2\n",
    "\n",
    "Boyer-Moore: How many alignments are skipped by the good suffix rule in this scenario?\n",
    "\n",
    "T: GGCTATAATGCGTA\n",
    "\n",
    "P:  TAATTAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6fa452ec-1688-4135-a1c0-1dc99391a70e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "skip = P_bm.good_suffix_rule(4)\n",
    "skipped_alignments = skip - 1\n",
    "print(skipped_alignments)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df0e5802-4e5d-4e53-9502-44ff883e5187",
   "metadata": {},
   "source": [
    "## Project"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e5e58ab-e4e1-4245-adb2-4a48d9fc1930",
   "metadata": {},
   "source": [
    "#### Question 1\n",
    "\n",
    "How many alignments does the naive exact matching algorithm try when matching the string \n",
    "GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG (derived from human Alu sequences) to the excerpt of human chromosome 1?  (Don't consider reverse complements.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "373782b4-8874-4074-9d15-19979b6954c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome=''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "genome = readGenome('chr1.GRCh38.excerpt.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "0ae866bc-4253-4007-ab0b-b0089c6de265",
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    alignments = 0\n",
    "    comparisons = 0\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        alignments += 1\n",
    "        match = True\n",
    "        for j in range(len(p)):\n",
    "            comparisons += 1\n",
    "            if t[i+j] != p[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)\n",
    "    return occurrences, alignments, comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d5b97a0a-2f3e-4308-9eaf-b442eb7d6a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "p = 'GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c50ce3ec-d7e9-4099-97d4-a5b016cfd9f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Naive alignments: 799954\n"
     ]
    }
   ],
   "source": [
    "num_alignments = len(genome) - len(p) + 1\n",
    "print(\"Naive alignments:\", num_alignments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "396fd095-a6f7-41f4-8bf4-01d848a3c1a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total alignments: 799954\n"
     ]
    }
   ],
   "source": [
    "# From another user\n",
    "occurrences, alignments, comparisons = naive('GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG', genome)\n",
    "print(f\"Total alignments: {alignments}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "167a5da3-35b8-4ef2-ae16-7b92e22efc55",
   "metadata": {},
   "source": [
    "#### Question 2\n",
    "\n",
    "How many character comparisons does the naive exact matching algorithm try when matching the string \n",
    "GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG (derived from human Alu sequences) to the excerpt of human chromosome 1?  (Don't consider reverse complements.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "085aabe1-033d-4910-b83b-99dc77a1421d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparisons done: 984143\n"
     ]
    }
   ],
   "source": [
    "print(\"Comparisons done:\", comparisons)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79738e6e-8a19-4fc4-90ee-e7d9d7b87ce0",
   "metadata": {},
   "source": [
    "#### Question 3\n",
    "\n",
    "How many alignments does Boyer-Moore try when matching the string \n",
    "GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG (derived from human Alu sequences) to the excerpt of human chromosome 1?  (Don't consider reverse complements.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f63b1c19-7bdb-4d18-9dd1-c1c9fec24182",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "127974\n"
     ]
    }
   ],
   "source": [
    "# occurrences, alignments, comparisons = boyer_moore('GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG', BoyerMoore('GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG'), genome)\n",
    "p_bm = BoyerMoore(p)\n",
    "occurrences, alignments, comparisons = boyer_moore(p, p_bm, genome)\n",
    "print(alignments)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a177ce75-3e86-4d21-9a35-a83e7236be10",
   "metadata": {},
   "source": [
    "#### Question 4\n",
    "\n",
    "Index-assisted approximate matching. In practicals, we built a Python class called Index implementing an ordered-list version of the k-mer index. The \n",
    "Index class is copied below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "206c8ff4-bdef-4f65-bd78-8dae783007bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def approximate_match_index(p, t, k):\n",
    "    segment_length = round(len(p) // (k+1))\n",
    "    hits = 0\n",
    "    all_matches = set()\n",
    "    index = Index(t, 8) # built on 8-mers\n",
    "    for i in range(k+1):\n",
    "        start = i * segment_length\n",
    "        end = min((i+1) * segment_length, len(p))\n",
    "        matches = index.query(p[start:end])\n",
    "        hits += len(matches)\n",
    "        for m in matches:\n",
    "            text_offset = m - start\n",
    "            if text_offset < 0 or (text_offset + len(p)) > len(t):\n",
    "                continue\n",
    "            mismatches = 0\n",
    "            for j in range(0, start):\n",
    "                if not p[j] == t[text_offset + j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > k:\n",
    "                        break\n",
    "            for j in range(end, len(p)):\n",
    "                if not p[j] == t[text_offset + j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > k:\n",
    "                        break\n",
    "            if mismatches <= k:\n",
    "                all_matches.add(text_offset)\n",
    "    return list(all_matches), hits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93070bbc-3df5-4c00-9b00-eed0e9fd848f",
   "metadata": {},
   "source": [
    "We also implemented the pigeonhole principle using Boyer-Moore as our exact matching algorithm.\n",
    "\n",
    "Implement the pigeonhole principle using Index to find exact matches for the partitions. Assume P always has length 24, and that we are looking for approximate matches with up to 2 mismatches (substitutions). We will use an 8-mer index.\n",
    "\n",
    "Write a function that, given a length-24 pattern P and given an Index object built on 8-mers, finds all approximate occurrences of P within T with up to 2 mismatches. Insertions and deletions are not allowed. Don't consider any reverse complements.\n",
    "\n",
    "How many times does the string GGCGCGGTGGCTCACGCCTGTAAT, which is derived from a human Alu sequence, occur with up to 2 substitutions in the excerpt of human chromosome 1?  (Don't consider reverse complements here.)\n",
    "\n",
    "Hint 1: Multiple index hits might direct you to the same match multiple times, but be careful not to count a match more than once.\n",
    "\n",
    "Hint 2: You can check your work by comparing the output of your new function to that of the naive_2mm function implemented in the previous module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "5e511d23-ab0a-411a-a608-6a55b78a4263",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "occurrences: 19\n"
     ]
    }
   ],
   "source": [
    "t = genome\n",
    "p = 'GGCGCGGTGGCTCACGCCTGTAAT'\n",
    "k = 2\n",
    "ans, hits = approximate_match_index(p, t, k)\n",
    "print(\"occurrences: \" + str(len(ans)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01fbbad3-0beb-408f-b2e6-550e420c3ef4",
   "metadata": {},
   "source": [
    "#### Question 5\n",
    "\n",
    "Using the instructions given in Question 4, how many total index hits are there when searching for occurrences of \n",
    "GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the excerpt of human chromosome 1?\n",
    "\n",
    "  (Don't consider reverse complements.)\n",
    "\n",
    "Hint: You should be able to use the boyer_moore function (or the slower naive function) to double-check your answer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "95e78f5c-05e9-49e1-a8a9-6c6cff5d26ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index hits: 90\n"
     ]
    }
   ],
   "source": [
    "print(\"index hits: \" + str(hits))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cee8b93a-3f55-474d-8025-2f07cb430539",
   "metadata": {},
   "source": [
    "#### Question 6\n",
    "Let's examine whether there is a benefit to using an index built using subsequences of T rather than substrings, as we discussed in the \"Variations on k-mer indexes\" video.  We'll consider subsequences involving every N characters.  For example, if we split ATATAT into two substring partitions, we would get partitions ATA (the first half) and TAT (second half).  But if we split ATATAT into two  subsequences  by taking every other character, we would get AAA(first, third and fifth characters) and TTT (second, fourth and sixth).\n",
    "\n",
    "Another way to visualize this is using numbers to show how each character of P is allocated to a partition.  Splitting a length-6 pattern into two substrings could be represented as 111222 and splitting into two subsequences of every other character could be represented as 121212\n",
    "\n",
    "The following class SubseqIndex is a more general implementation of Index that additionally handles subsequences. It only considers subsequences that take every Nth character:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ba1a8136-675e-49b0-8187-a96d2451a278",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bisect\n",
    "   \n",
    "class SubseqIndex(object):\n",
    "    \"\"\" Holds a subsequence index for a text T \"\"\"\n",
    "    \n",
    "    def __init__(self, t, k, ival):\n",
    "        \"\"\" Create index from all subsequences consisting of k characters\n",
    "            spaced ival positions apart.  E.g., SubseqIndex(\"ATAT\", 2, 2)\n",
    "            extracts (\"AA\", 0) and (\"TT\", 1). \"\"\"\n",
    "        self.k = k  # num characters per subsequence extracted\n",
    "        self.ival = ival  # space between them; 1=adjacent, 2=every other, etc\n",
    "        self.index = []\n",
    "        self.span = 1 + ival * (k - 1)\n",
    "        for i in range(len(t) - self.span + 1):  # for each subseq\n",
    "            self.index.append((t[i:i+self.span:ival], i))  # add (subseq, offset)\n",
    "        self.index.sort()  # alphabetize by subseq\n",
    "    \n",
    "    def query(self, p):\n",
    "        \"\"\" Return index hits for first subseq of p \"\"\"\n",
    "        subseq = p[:self.span:self.ival]  # query with first subseq\n",
    "        i = bisect.bisect_left(self.index, (subseq, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != subseq:\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "927805a2-f4b4-40e7-a28b-86861b00553c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('AAA', 0), ('TTT', 1)]\n"
     ]
    }
   ],
   "source": [
    "# For example, if we do:\n",
    "ind = SubseqIndex('ATATAT', 3, 2)\n",
    "print(ind.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "7a79d57f-da22-4054-a5b0-e9ef57341722",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "# And if we query this index:\n",
    "p = 'TTATAT'\n",
    "print(ind.query(p[0:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "cff931cc-9422-4f47-9131-bb040e311655",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1]\n"
     ]
    }
   ],
   "source": [
    "# because the subsequence TAA is not in the index. But if we query with the second subsequence:\n",
    "print(ind.query(p[1:]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96cf4d02-688c-45cd-8f1f-38778b5302d4",
   "metadata": {},
   "source": [
    "Write a function that, given a length-24 pattern P and given a SubseqIndex object built with k = 8 and ival = 3, finds all approximate occurrences of P within T with up to 2 mismatches.\n",
    "\n",
    "When using this function, how many total index hits are there when searching for GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the excerpt of human chromosome 1?  (Again, don't consider reverse complements.)\n",
    "\n",
    "Hint: See this notebook for a few examples you can use to test your function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "fe9a4399-df7e-4581-9dcc-cddd7e1e3ac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def approximate_match_subseq_index(p, t, k):\n",
    "    segment_length = round(len(p) // (k+1))\n",
    "    hits = 0\n",
    "    all_matches = set()\n",
    "    index = SubseqIndex(t, 8, 3) # built on 8-mers and subsequence intervals of 3\n",
    "    for start in range(k+1):\n",
    "        matches = index.query(p[start:])\n",
    "        hits += len(matches)\n",
    "        for m in matches:\n",
    "            text_offset = m - start\n",
    "            if text_offset < 0 or (text_offset + len(p)) > len(t):\n",
    "                continue\n",
    "            mismatches = 0\n",
    "            for j in range(0, len(p)):\n",
    "                if not p[j] == t[text_offset + j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > k:\n",
    "                        break\n",
    "            if mismatches <= k:\n",
    "                all_matches.add(text_offset)\n",
    "    return list(all_matches), hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "c95cb071-6bd9-4e79-8fd3-e8613cf5ae42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "79\n"
     ]
    }
   ],
   "source": [
    "t = genome\n",
    "p = 'GGCGCGGTGGCTCACGCCTGTAAT'\n",
    "k = 2\n",
    "_, hits = approximate_match_subseq_index(p, t, k)\n",
    "print(hits)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
